### PR TITLE
release-tool: select-test-suite command

### PR DESCRIPTION
**Pull request by mchalski**
https://tracker.mender.io/browse/MEN-2660

the command will be used right before running integration tests to determine which test suite to run - open, enterprise, or both.

it reaches out into git repositories to check which components are not in recent master version. this is taken to mean that we're e.g. on a PR, or on a custom revision in general (i.e. this component 'changed').

more specifically, we're checking backend services versions, because they can potentially restrict the test suite to 'just open' or 'just enterprise' tests. 

the rules for test selection are slightly tricky:
- if we're building strictly open services - we must run all tests. open services occur in both open and enterprise setups
- if we're building strictly enteprise services - of course those appear only in the enterprise setup
- if we're building open services, but they have their enteprise counterparts (deployments, useradm,...), we can safely run just 'open' tests. 

there are some corner cases too:
- we might not be building any services at all. maybe we're building mender-cli, or the gateway - make no assumptions and just run everything then

one practical remark:
I could have used the existing component maps to indicate - which components are 'services', which are open, open-with-enterprise, or enterprise.
instead I opted for separate lists (sets) specifying those services verbatim. I didn't want to pollute the existing structs, this way it's clear these entities have different purposes. the component maps are still used strictly for mapping git to docker images/containers.

### CODE REVIEW

File: extra/release_tool.py [hunk 1]
```diff
@@ -2133,6 +2140,68 @@ def do_verify_integration_references(args, optional_too):
         print("\nMake sure all *.yml files contain the correct versions.")
         sys.exit(1)
 
+def is_repo_fresh_master(path):
+    """ Check if we're on the most recent commit in 'master'.
+    """
+    sha_master = execute_git(None, path, ["rev-parse", "origin/master"], capture=True, capture_stderr=True)
```
Comment (lluiscampos) [inline extra/release_tool.py]: You should not hardcode "origin" here, there are developers (like me) that use a different remote name. See code using `find_upstream_remote(state, repo.git())` for inspiration, for instance [here](https://github.com/mendersoftware/integration/blob/master/extra/release_tool.py#L1431).
```

Comment (lluiscampos) [reply extra/release_tool.py]: Sorry, published this early by mistake! I will prepare the rest of the comments in one go, you can wait for the rework.

File: extra/release_tool.py [hunk 2]
```diff
@@ -2133,6 +2140,75 @@ def do_verify_integration_references(args, optional_too):
...
+    # if we're building only backend services - we can proceed with test selection
+    # if not - assume we must run all test suites (e.g. for mender-cli, etc.)
+    non_service_components = built_components - BACKEND_SERVICES
+    if len(non_service_components) > 0:
+        return "all"
+
+    built_services = BACKEND_SERVICES & built_components
+
+    # no services are being built in custom revs - just run all tests
+    if len(built_services) == 0:
```
[the remark this answers is not in the document]
Comment (mchalski) [reply extra/release_tool.py]: you're right if there are built components, and none of them are non-service, then it's just >1 service. will fix.

File: extra/release_tool.py [hunk 3]
```diff
@@ -2133,6 +2140,75 @@ def do_verify_integration_references(args, optional_too):
...
+        - open services with closed versions should trigger the 'open' test suite
+        - enterprise services can run just the 'enterprise' setup
+    """
+    # check all known git components for custom revisions
+    # answers the question what we're actually building
+    built_components = set({})
+    for repo in Component.get_components_of_type("git", only_release=False):
+        path = os.path.abspath(os.path.join(git_checkout, repo.git()))
+        if not os.path.exists(path):
+            continue
```
[the remark this answers is not in the document]
Comment (mchalski) [reply extra/release_tool.py]: right - it's a shortcut I made for manual debugging (I don't have *all* those clones locally).
will fix.

File: extra/release_tool.py [hunk 4]
```diff
@@ -2133,6 +2140,75 @@ def do_verify_integration_references(args, optional_too):
...
+        Rules:
+        - open services, without closed versions, should trigger both setup test runs
+        - open services with closed versions should trigger the 'open' test suite
+        - enterprise services can run just the 'enterprise' setup
+    """
+    # check all known git components for custom revisions
+    # answers the question what we're actually building
+    built_components = set({})
+    for repo in Component.get_components_of_type("git", only_release=False):
+        path = os.path.abspath(os.path.join(git_checkout, repo.git()))
```
[the remark this answers is not in the document]
Comment (mchalski) [reply extra/release_tool.py]: Can you elaborate? I thought component maps include everything we have, what do you mean by 'go code'?

File: extra/release_tool.py (hunk 1 above)
Comment (mchalski) [reply extra/release_tool.py]: fixed, can you check?

File: extra/release_tool.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mchalski) [reply extra/release_tool.py]: fixed, can you check?

File: extra/release_tool.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mchalski) [reply extra/release_tool.py]: fixed, can you check?

**Comment by mchalski** [comment]
@lluiscampos I need clarification on `However, I think there is something missing: this will not work for PR over releases`.

do you mean - PRs to older branches, as in - a backported bugfix?

File: extra/release_tool.py (hunk 1 above)
Comment (mchalski) [reply extra/release_tool.py]: ps. I might still do some fixes concerning PRs for releases that you mentioned, for now it's a basic fix.

File: extra/release_tool.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lluiscampos) [reply extra/release_tool.py]: Let me try to elaborate.

So "git_checkout" here will be "..", so you expect all git repos to be there. However, `go` requires the full path `go/src/mendersoftware...`, and a common way to do this (and the way our CIs do it) is to get them in `../go/src...` from where integration lives.

See here: https://github.com/mendersoftware/integration/blob/master/extra/release_tool.py#L2097

[the remark this answers is not in the document]
Comment (lluiscampos) [reply extra/release_tool.py]: In my local setup I have the repos in ".." and then symlinked to `home/lluis/go/src/...`

**Comment by lluiscampos** [comment]
> @lluiscampos I need clarification on `However, I think there is something missing: this will not work for PR over releases`.
> 
> do you mean - PRs to older branches, as in - a backported bugfix?

Yes, that I meant.

File: extra/release_tool.py (hunk 1 above)
Comment (lluiscampos) [reply extra/release_tool.py]: :+1:

File: extra/release_tool.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lluiscampos) [reply extra/release_tool.py]: :+1:

File: extra/release_tool.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mchalski) [reply extra/release_tool.py]: so would the fix be better in this place? https://github.com/mendersoftware/integration/pull/669/files#diff-b1f07a2050f82dea7314df2fbfc4053eR2205

I'm still not sure if I get this. Will this work?:

- factor this bit into common function:
https://github.com/mendersoftware/integration/pull/669/files#diff-b1f07a2050f82dea7314df2fbfc4053eR2205
- use in both locations - original and in do_select_test_suite?

**Comment by mchalski** [comment]
@lluiscampos I propose we drop the PR over release thing. Maybe someone more competent can take it over as an improvement, I feel I'm out of my depth here and this task blew up already. Is that ok?

I will fix the path selection though.

File: extra/release_tool.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mchalski) [reply extra/release_tool.py]: ahh, nevermind. checking the main git checkout dir is useless - doesn't guarantee any repos are in there. so it has to be done on a per-repo basis. hold on, will fix.

**Comment by lluiscampos** [comment]
> @lluiscampos I propose we drop the PR over release thing. Maybe someone more competent can take it over as an improvement, I feel I'm out of my depth here and this task blew up already. Is that ok?
> 
> I will fix the path selection though.

Sounds reasonable :+1:

File: extra/release_tool.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lluiscampos) [reply extra/release_tool.py]: Yes, not all repos are "go code", so the might or might not be in the same parent repo.

[the remark this answers is not in the document]
Comment (mchalski) [reply extra/release_tool.py]: submitted a fix - it's a separate commit to make it stand out clearly.

File: extra/release_tool.py [hunk 5]
```diff
@@ -2079,6 +2086,17 @@ def figure_out_checked_out_revision(state, repo_git):
 
     return [(ref, "tag") for ref in refs]
 
+def find_repo_path(name, paths):
+    """ Try to find the git repo 'name' under some known paths.
```
Comment (lluiscampos) [inline extra/release_tool.py]: Rather than "known paths" it tries to figure out which of the "given paths" do exist.

Actually, this has nothing to do with repos. What about something like `find_existing_path(dir_name, parent_candidates)`.

Comment (lluiscampos) [reply extra/release_tool.py]: This is completely low priority.
